icicle-compiler/src/Icicle/Runtime/Logical.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# OPTIONS_GHC -funbox-strict-fields #-}
+module Icicle.Runtime.Logical (
+    Value(..)
+  ) where
+
+import           Data.ByteString (ByteString)
+import           Data.Map (Map)
+import qualified Data.Vector as Boxed
+
+import           GHC.Generics (Generic)
+
+import           Icicle.Runtime.Data
+
+import           P
+
+
+data Value =
+    Unit
+  | Bool !Bool64
+  | Int !Int64
+  | Double !Double
+  | Time !Time64
+
+  | Left !Value
+  | Right !Value
+
+  | None
+  | Some !Value
+
+  | Error !Error64
+  | Success !Value
```

Comment:
if `Success` means `Right :: a -> Sum Error a`, you probably want another constructor to distinguish between `Left Error` and raw, unadorned `Error`. unadorned errors are unlikely to occur, but I imagine we can encode them?